libsolidity/codegen/ContractCompiler.cpp

```diff
@@ -691,7 +687,13 @@ bool ContractCompiler::visit(WhileStatement const& _whileStatement)
 		m_context.appendConditionalJumpTo(loopEnd);
 	}
 
+	solAssert(m_loops.top() == &_whileStatement, "");
+	m_loops.pop();
+
 	m_context.appendJumpTo(loopStart);
```

Comment:
I kind of have the impression that this should move above the `solAssert`, although that is subjective.